### PR TITLE
Show context menu for playlist and its items

### PR DESCRIPTION
**Pull request by sangwoo108**
Add menu buttons for playlist and its items.
When clicking them, a related contextual menu should show up.

Currently, Leo components are not updated yet, some icons and colors are out of sync, but hopefully it'll be sorted out by the Leo update.

### From storybook
image

image

Resolves https://github.com/brave/brave-browser/issues/31969

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/playlist/browser/resources/components/header.tsx [hunk 1]
```diff
@@ -64,14 +69,36 @@ export default function Header ({ playlistId }: HeaderProps) {
...
           />
+          <ContextualMenuAnchorButton
+            items={[
+              { name: 'Rename', iconName: 'edit-box', onClick: () => {} },
+              {
+                name: 'Edit',
+                iconName: 'list-bullet-default',
+                onClick: () => {}
+              },
+              { name: 'Share', iconName: 'share', onClick: () => {} },
```
Comment (sangwoo108) [inline components/playlist/browser/resources/components/header.tsx:88]: I'm curious if this 'share' icon could be interpreted as 'share-macos' and 'share-windows' depending on the platform. Actually, our design guide says this icon should be share-macos.

File: components/playlist/browser/resources/components/contextualMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  display: grid;
+  grid-template-columns: 1fr 16px;
+  grid-auto-rows: 28px;
+  width: fit-content;
+  gap: 0px 8px;
+  align-items: center;
+
+  padding: 8px 16px;
+  border-radius: 8px;
+  border: 1px solid var(--leo-color-gray-20);
```
Comment (fallaciousreasoning) [inline components/playlist/browser/resources/components/contextualMenu.tsx:68]: Out of curiosity, is there a reason this one doesn't use an interpolation like below?

File: components/playlist/browser/resources/components/contextualMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+const StyledMenuIcon = styled(Icon)`
+  --leo-icon-size: 16px;
+`
+
+function ContextualMenuItem ({
+  name,
+  iconName,
+  onClick,
+  onDismiss
+}: MenuItemProps & { onDismiss: () => void }) {
```
Comment (fallaciousreasoning) [inline components/playlist/browser/resources/components/contextualMenu.tsx:41]: Any reason `onDismiss` is separate from the other props here?

File: components/playlist/browser/resources/components/contextualMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  padding: 8px 16px;
+  border-radius: 8px;
+  border: 1px solid var(--leo-color-gray-20);
+  background: ${color.white};
+`
+
+function ContextualMenu ({
+  items,
+  onDismiss
+}: MenuProps & { onDismiss: () => void }) {
```
Comment (fallaciousreasoning) [inline components/playlist/browser/resources/components/contextualMenu.tsx:75]: `MenuProps` has an optional `onDismissMenu` prop - is that different to this one?

File: components/playlist/browser/resources/components/contextualMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  React.useEffect(() => {
+    document.addEventListener('click', onDismiss)
+    // Playlist has an <iframe> that holds video player. When it's clicked,
+    // we should dismiss the menu but 'click' event won't be notified. Instead
+    // of it, uses 'blur' event while it could have a little side effect.
+    window.addEventListener('blur', onDismiss)
```
Comment (fallaciousreasoning) [inline components/playlist/browser/resources/components/contextualMenu.tsx:81]: :+1:

File: components/playlist/browser/resources/components/header.tsx (hunk 1 above)
Comment (fallaciousreasoning) [reply components/playlist/browser/resources/components/header.tsx:88]: It won't changed based on platform, nah - do you reckon it'd be a useful thing to add to Leo?

File: components/playlist/browser/resources/components/playerSeeker.tsx [hunk 6]
```diff
@@ -200,7 +200,7 @@ export default function PlayerSeeker ({ videoElement }: Props) {
 
   React.useEffect(() => {
     if (!videoElement) {
-      console.error(`Video element doesn't exist`)
+      // Can be null on start up.
```
Comment (fallaciousreasoning) [inline components/playlist/browser/resources/components/playerSeeker.tsx:203]: Ah yeah :/ I missed that last time sorry :laugh

File: components/playlist/browser/resources/stories/components.tsx [hunk 7]
```diff
@@ -58,3 +59,28 @@ export const VideoPlayer = () => {
...
+        items={[
+          {
+            name: 'test',
+            iconName: 'trash',
+            onClick: () => {
+              window.alert('test')
+            }
+          },
+          {
+            name: 'looooooooooooooooooooong text this is really long',
```
Comment (fallaciousreasoning) [inline components/playlist/browser/resources/stories/components.tsx:76]: :laughing:

**Review by fallaciousreasoning — APPROVED** [review]
LGTM

File: components/playlist/browser/resources/components/contextualMenu.tsx (hunk 4 above)
Comment (sangwoo108) [reply components/playlist/browser/resources/components/contextualMenu.tsx]: Ah, yeah, `onDismiss` here is more like 'onClickOutside' for weaving ContextualMenu and AnchorButton. I didn't want to expose this to clients, so separated from the MenuProps. same [for sko/pl-menu](https://github.com/brave/brave-core/pull/19482/files#r1280104526)

I feel like it'd be better to rename this to `onClickOutside`

File: components/playlist/browser/resources/components/header.tsx (hunk 1 above)
Comment (sangwoo108) [reply components/playlist/browser/resources/components/header.tsx:88]: I'm not sure, but depending on what @aguscruiz was intended. But now I think I should rever this back to 'share-macos' 😄

@aguscruiz , when you use 'share-macos' icon, do you mean to use 'share-macos' across all platform, or use platform specific icon?
If it's the latter, I would automatically interpreted Leo icon would be nice. It's interesting that we don't have 'share-linux' icon though.

File: components/playlist/browser/resources/components/contextualMenu.tsx (hunk 2 above)
Comment (sangwoo108) [reply components/playlist/browser/resources/components/contextualMenu.tsx]: Oops, my bad. Didn't notice that we have gray[20] for this 😃

File: components/playlist/browser/resources/stories/components.tsx (hunk 7 above)
Comment (sangwoo108) [reply components/playlist/browser/resources/stories/components.tsx:76]: 🤭

File: components/playlist/browser/resources/components/playerSeeker.tsx (hunk 6 above)
Comment (sangwoo108) [reply components/playlist/browser/resources/components/playerSeeker.tsx:203]: No worries at all 😄

**Pull request merged**